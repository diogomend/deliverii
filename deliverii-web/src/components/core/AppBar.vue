<template>
  <v-app-bar id="app-bar" absolute app color="transparent" flat height="75">
    <v-btn
      class="mr-3"
      elevation="1"
      fab
      small
      @click="toggleDrawer(!drawer)"
      v-if="isMobile && isLoggedIn"
    >
      <v-icon v-if="value">
        mdi-view-quilt
      </v-icon>

      <v-icon v-else>
        mdi-dots-vertical
      </v-icon>
    </v-btn>

    <v-toolbar-title
      class="hidden-sm-and-down font-weight-light"
      v-text="$route.name"
    />

    <v-spacer />

    <v-btn v-if="isLoggedIn" @click="doLogout">
      <v-icon class="mr-5">mdi-account</v-icon>
      Logout
    </v-btn>
  </v-app-bar>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import MobileMixin from "@/mixins/mobileMixin";
export default {
  name: "AppBar",
  props: {
    value: {
      type: Boolean,
      default: false
    }
  },
  mixins: [MobileMixin],

  computed: {
    ...mapGetters({
      drawer: "appState/drawer",
      isLoggedIn: "auth/isLoggedIn"
    })
  },

  methods: {
    ...mapActions("appState", ["toggleDrawer"]),
    ...mapActions("auth", ["logout"]),
    doLogout() {
      this.logout();
      this.$router.push("/login");
    }
  }
};
</script>
