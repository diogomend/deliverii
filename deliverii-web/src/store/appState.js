const getTimeout = alert => {
  return "timeout" in alert ? alert.timeout : 5000;
};

const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  },
  addAlert: ({ commit, getters }, alert) => {
    const timeout = getTimeout(alert);
    const alerts = getters.getAlerts;
    const repeatedIndex = alerts.findIndex(val => {
      return val.type === alert.type && val.message === alert.message;
    });
    let id;

    do {
      id = Date.now() + Math.random();
    } while (getters.getAlerts.some(elem => elem.id === id));

    commit("ADD_ALERT", { ...alert, id });

    if (repeatedIndex !== -1) {
      commit("REMOVE_ALERT", repeatedIndex);
    }

    setTimeout(() => {
      const index = getters.getAlerts.findIndex(elem => elem.id === id);

      if (index !== -1) {
        commit("REMOVE_ALERT", index);
      }
    }, timeout);
  }
};

const mutations = {
  ADD_ALERT(state, { type, message, id }) {
    state.alert.push({
      type,
      message,
      id
    });
  },
  REMOVE_ALERT(state, index) {
    state.alert.splice(index, 1);
  },
  TOGGLE_LOADING(state, loading) {
    state.isLoadingSession = loading;
  },
  TOGGLE_DRAWER(state, drawer) {
    state.drawer = drawer;
  }
};

const getters = {
  getAlerts: state => state.alert,
  getLoadingSession: state => state.isLoadingSession,
  drawer: state => state.drawer
};

const state = () => ({
  drawer: false,
  isLoadingSession: true,
  alert: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
