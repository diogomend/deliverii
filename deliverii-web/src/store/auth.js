import auth from "../service/auth";
const actions = {
  login: async ({ commit, dispatch }, { email, password }) => {
    const res = await auth.login(email, password);
    if (res) {
      commit("SET_USER", res.user);
      commit("SET_TOKEN", res.token);
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  register: async ({ dispatch }, registerObj)=> {
    const res = await auth.register(registerObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Registered Successfully`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  logout({ commit }) {
    commit("SET_USER", false);
    commit("SET_TOKEN", false);

    auth.logoutUser();
  },

  async loadSession({ commit }) {
    const isLoggedIn = auth.isLoggedIn();
    if (isLoggedIn) {
      const user = auth.getUserInfo();
      commit("SET_USER", user);
    }
  },
};

const mutations = {
  SET_USER(state, user) {
    state.user = user;
  },
  SET_TOKEN(state, token) {
    state.token = token;
  }
};

const getters = {
  getUser: state => state.user,
  isLoggedIn: state => !!state.user
};

const state = () => ({
  user: false,
  token: false
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
