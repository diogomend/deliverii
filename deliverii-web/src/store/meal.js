import meal from "../service/meal";

const actions = {
  fetchMeals: async (_, restaurantID) => {
    return await meal.fetchMeals(restaurantID);
  },
  fetchMeal: async (_, { restaurant, meal: mealID }) => {
    const mealList = await meal.fetchMeal(restaurant, mealID);
    if (mealList && mealList.length) {
      return mealList[0];
    }

    return false;
  },

  updateMeal: async ({ dispatch }, { restaurant, mealID, mealObj }) => {
    const res = await meal.updateMeal(restaurant, mealID, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  },

  createMeal: async ({ dispatch }, { restaurant, mealObj }) => {
    const res = await meal.createMeal(restaurant, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  }
};

const mutations = {};

const getters = {};

const state = () => ({
  meals: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
