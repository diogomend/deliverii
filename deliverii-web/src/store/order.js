import order from "../service/order";

const actions = {
  fetchAll: async ({ commit }, id) => {
    const res = await order.fetchAll(id);
    commit("ADD_ORDERS", res);
  },
  fetchOrder: async (_, id) => {
    return await order.fetchOrder(id);
  },
  updateStatus: async (_, { orderID, status }) => {
    return await order.updateStatus(orderID, status);
  },
  loadCart: async ({ commit }) => {
    const cart = await order.loadCart();
    if (cart) {
      commit("ADD_CART", JSON.parse(cart));
    }
  },
  addCart({ commit }, cart) {
    order.saveCart(cart);
    commit("ADD_CART", cart);
  },
  addAddress({ commit }, address) {
    commit("ADD_ADDRESS", address);
  },
  async makeOrder({ dispatch, getters }) {
    const cartObj = getters.getCart;
    const meals = cartObj.cart.map(item => {
      return {
        meal: item._id,
        quantity: item.quantity
      };
    });

    const request = {
      meals: meals,
      restaurantID: cartObj.restaurantID,
      address: cartObj.address
    };

    const ret = await order.makeOrder(request);
    if (ret._id) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Order created successfully`
        },
        { root: true }
      );
      return ret._id;
    }

    return false;
  }
};

const mutations = {
  ADD_ORDERS(state, orders) {
    state.orders = orders;
  },
  ADD_CART(state, cart) {
    state.cart = cart;
  },
  ADD_ADDRESS(state, address) {
    state.cart.address = address;
  }
};

const getters = {
  getOrders: state => state.orders,
  getCart: state => state.cart
};

const state = () => ({
  orders: [],
  cart: {}
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
