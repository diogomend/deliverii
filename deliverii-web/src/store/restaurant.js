import restaurant from "../service/restaurant";

const actions = {
  fetchAll: async ({ commit }) => {
    const res = await restaurant.fetchAll();
    commit("ADD_RESTAURANTS", res);
  },
  fetchRestaurantInfo: async (_, id) => {
    return await restaurant.fetchRestaurant(id);
  },
  deleteRestaurant: async ({ dispatch }, id) => {
    const res = await restaurant.deleteRestaurant(id);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Delete successfully` },
        { root: true }
      );
    }

    return res;
  },
  createRestaurant: async ({ dispatch }, restaurantObj) => {
    const res = await restaurant.createRestaurant(restaurantObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Created successfully` },
        { root: true }
      );
    }

    return res;
  },

  addBlacklist: async ({ dispatch }, { restaurantID, customerID }) => {
    const res = await restaurant.addBlacklist(restaurantID, customerID);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Added to blacklist` },
        { root: true }
      );
    }

    return res;
  }
};

const mutations = {
  ADD_RESTAURANTS(state, restaurants) {
    state.restaurants = restaurants;
  }
};

const getters = {
  getRestaurants: state => state.restaurants
};

const state = () => ({
  restaurants: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
