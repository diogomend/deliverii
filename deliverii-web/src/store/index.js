import Vue from "vue";
import Vuex from "vuex";
import appState from "@/store/appState";
import auth from "@/store/auth";
import order from "@/store/order";
import restaurant from "@/store/restaurant";
import meal from "@/store/meal";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    appState,
    auth,
    order,
    restaurant,
    meal
  }
};

export default new Vuex.Store(storeOptions);
