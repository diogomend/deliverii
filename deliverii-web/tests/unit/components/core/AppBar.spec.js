import Vuetify from 'vuetify';
import Vue from 'vue';
import Vuex from "vuex";
import AppBar from "@/components/core/AppBar.vue";
import { shallowMount } from "@vue/test-utils"
import { storeOptions } from "@/store/index"

Vue.use(Vuex);
Vue.use(Vuetify);


describe("AppBar.vue", () => {
    let store;

    beforeEach(() => {
        jest.clearAllMocks();
        store = new Vuex.Store(storeOptions);
      });

      it("is a Vue instance", () => {
        const wrapper = shallowMount(AppBar, {
          store,
          mocks: {
              $route: {
                  name: "MOCK_NAME"
              }
          }
        });
    
        const template = wrapper.html();
        expect(template).toMatchSnapshot();
      });

      it("is a Vue instance", () => {
        const wrapper = shallowMount(AppBar, {
          store,
          mocks: {
              $route: {
                  name: "MOCK_NAME"
              }
          }
        });
    
        const template = wrapper.html();
        expect(template).toMatchSnapshot();
      });

      it("is loggedIn should see logout-btn and trigger event", () => {
        const mockFn = jest.fn();
        const wrapper = shallowMount(AppBar, {
          store,
          mocks: {
              $route: {
                  name: "MOCK_NAME",
              },
              $router: {
                push: mockFn
              }
          },
          computed: {
            isLoggedIn() {
              return true
            }
          }
        });
        
        expect(wrapper.find('.logout-btn')).toBeTruthy();
        wrapper.vm.doLogout();
        expect(mockFn).toHaveBeenCalled();
      });
})