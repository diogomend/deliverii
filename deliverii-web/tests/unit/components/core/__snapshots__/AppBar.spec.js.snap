// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar.vue is a Vue instance 1`] = `
<v-app-bar-stub color="transparent" height="75" tag="header" tile="true" absolute="true" extensionheight="48" flat="true" src="" value="true" app="true" id="app-bar">
  <!---->
  <v-toolbar-title-stub class="hidden-sm-and-down font-weight-light"></v-toolbar-title-stub>
  <v-spacer-stub></v-spacer-stub>
  <!---->
</v-app-bar-stub>
`;

exports[`AppBar.vue is a Vue instance 2`] = `
<v-app-bar-stub color="transparent" height="75" tag="header" tile="true" absolute="true" extensionheight="48" flat="true" src="" value="true" app="true" id="app-bar">
  <!---->
  <v-toolbar-title-stub class="hidden-sm-and-down font-weight-light"></v-toolbar-title-stub>
  <v-spacer-stub></v-spacer-stub>
  <!---->
</v-app-bar-stub>
`;
